magento/pwa-studio
# Introduce Targetables and expand target coverage

## Description

### Features
- Add [Targetables](https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md), a new helper library for powerful Targets
- Reimplement core Targets using Targetables
- Add an automatic API generator to Peregrine that exposes all hooks and talons in Targets

Excerpt from the explainer at https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md follows:

> # Targetables: Fluent tools for making your code extensible with Targets
> 
> Targetables are handy objects which represent source files in your project or library. When paired with Buildpack Targets, Targetables make it simple to customize PWA modules on the fly.
> 
> ## Introduction
> 
> You need a lot of Targets to make a PWA Studio project powerfully extensible. Enhancing the storefront through Targets means changing source code in highly specific ways during the build process. A Targetable for a source file provides methods for making pinpoint changes to the code in that file.
> 
> ```js
> targetables.esModule('@my/library/Button.js', {
>   publish() {
>     const focusClasses = this
>       .addImport('focusClasses from "src/a11y/focus-classes.css"');
> 
>     this
>       .setJSXProps('<button>', { tabIndex: '0' })
>       .addJSXClassName('<button>', `${focusClasses}.focusRing`)
>   }
> })
> ```
> 
> The above modifies the source of `Button.js` right before Webpack pulls it into the build. It's always on the fly, never saving those changes to `Button.js` on disk.

### Other changes

- Moved `ReactRefreshWebpackPlugin` to PWADevServer, so it only runs in watch mode. A plain build in dev mode caused an error.
- Switched from using `babel-loader` to a simple custom Babel loader. That loader adds the ability for Babel plugins to report errors and warnings that Webpack will log. Otherwise it works exactly the same.
- Fixed a bad idea in BuildBusPlugin, which was to queue log messages. Now it just logs them immediately.
- Moved the default routes in `venia-ui` out of the intercept itself and into a `defaultRoutes.json` file.
- Other very small bugfixes I discovered along the way.

## Related Issue





Closes PWA-875.

## Acceptance 



### Verification Stakeholders
The whole @magento/pwa-studio-team 
Community leaders @larsroettig and @Jordaneisenburger 
Community participants
### Specification



### Verification Steps


1. Go to the [Demo section of the explainer](https://github.com/magento/pwa-studio/blob/zetlen/targetables/pwa-devdocs/_drafts/targetables-explainer.md#demo). 
1. Copy and paste the code there into `venia-concept/local-intercept.js` as instructed.
1. Run `yarn watch:venia`
1. Load the Venia site in your browser.
1. Observe the changes made by Targetables.
1. Repeat for production mode.

## Screenshots / Screen Captures (if appropriate)
![image](https://user-images.githubusercontent.com/1643758/95467180-c7d44180-0942-11eb-8d5d-2e6f43e79821.png)

Look, I was playing around. It works, okay?

## Checklist

* I have updated the documentation accordingly, if necessary.

## Notes

- The HookInterceptorSet reads the filesystem to build Peregrine APIs. This means that if someone adds, removes, or moves a while, those snapshots will break. This is on purpose; it tells the developer that changes have occurred in the public API! It may be necessary to modify or subclass HookInterceptorSet for compatibility in the future.

- @jcalcaben will need to decide how to document the talon and hook targets that are autogenerated.

- TargetableSet will probably need more and more utility methods. Arguably they don't go there, but it works today.

- I tried making the JSX methods in TargetableReactComponent more general-purpose and powerful by using [Babel templates](https://babeljs.io/docs/en/next/babel-template.html). Two main things went wrong:

  - Their API requires a lot of introspection, and doesn't tolerate any unknown properties.
  - Syntactic placeholders aren't compatible with being in JSXText position, so they broke in nonintuitive ways.

  This is worth more investigation by someone. Babel templates, or even callbacks, would make it much easier to create new operations. As it is, the `BabelModifyJSXPlugin` is pretty large, and each common method has to be made separately.

- I recommend that someone follow this up with a TargetableCSSModule that lets you do PostCSS transforms, to demonstrate that Targetables aren't just for JS.
